Feature/onboarding

dalbanhi opened the pull request:
## New Features Implemented
- [x] implemented user onboarding
- [x] routing from sign-up to onboarding
- [x] validation on the back end and front end with zod

dalbanhi committed:
added hasOnbarded to model
:

dalbanhi committed:
barebones onboarding flow with react hook forms set up

dalbanhi committed:
added image metadata and keeping track of uploaded imageurl in step 1

dalbanhi committed:
finished step 2

dalbanhi committed:
step 3 done

dalbanhi committed:
added step 4, need to do form resolver thing

dalbanhi committed:
added step 4 and resolution and server action

dalbanhi committed:
fixed reading of images from user

brandonetter commented:
I know you said that you wanted a shot at handling the redirect.

So I'll give you a shot at that before we get this merged, reach out if you have any questions. 

Let's also make sure to share the ENV with me so I can be on the same version as you (for the image upload) 👍

If you don't know where to start- consider checking out the docs for redirecting after sign-in!
https://next-auth.js.org/getting-started/client#signin

dalbanhi committed:
added dynamic import of creatable select from react-select to avoid warning

dalbanhi committed:
fixed error messages and routing after onboarding

dalbanhi commented:
Hey Brandon, I fixed the onboarding flow. Checkout this video here: [Loom video](https://www.loom.com/share/01a583dc430a4d8891062684d7dcd9b4?sid=380b53f6-452a-4167-a5ba-e8b81e5ba394)

brandonetter commented on components/shared/Form/Onboarding/OnboardingForm.tsx:
```diff
@@ -200,34 +214,62 @@ const OnboardingForm: React.FC<OnboardingFormProps> = ({ step, session }) => {
-    console.log(data);
+  const onSubmit = async (data: any) => {
     updateUser(data);
+    const session = await getSession();
```
Not sure we need this 'getSession'.

We could also await the update of the user before pushing them to the homepage:
```js
  const onSubmit = async (data: any) => {
    await updateUser(data);
    router.push("/");
  };
  ```

dalbanhi replied:
I thought I needed this in order to 'refresh' the session because I was having some issues with the rerouting. I see how awaiting for the updateUser to finish makes sense.

brandonetter commented on components/shared/Form/Onboarding/OnboardingForm.tsx:
On line 111, we can remove some of these unused variables from our useFieldArray

brandonetter reviewed (changes requested):
This feels like it works great- I do have a question about enforcing the square image though.  Is there a particular reason you chose to do that?

Your "TechStackSelector.tsx" is also causing a warning:
![image](https://github.com/dalbanhi/git-note/assets/4108484/17876bf8-6992-4b90-a7ad-3681a5a5a86a)

This is because we're spreading field that has a useRef:
```js
<CreatableSelect
          isMulti
          instanceId={"techStack"}
          {...field}
 ```
 
 You've already handled the "onChange", maybe we don't need this?
 
 This has also gotten quite large. We may want to make a `Steps.tsx` file or something similar to handle all the individually rendered steps. That way we still have a single component handling all the form logic- but the rendered steps can all be in one place away from the component.
 
 
 ```js
 // Steps.tsx
 "use client";
import React from "react";
import { useFieldArray, Control, Controller } from "react-hook-form";
import { Session } from "next-auth";

import Button from "~/components/ui/Button";
import Input from "@components/ui/Input";
import DeletableListItem from "@components/ui/DeletableListItem";
import ImageUploader from "./ImageUploader";
import { UseFormRegister, FieldValues } from "react-hook-form";
import TechStackSelector from "./TechStackSelector";
import MyDatePicker from "./MyDatePicker/MyDatePicker";

interface StepProps {
  register: UseFormRegister<FieldValues>;
  control?: Control<FieldValues>;
  session?: Session | null;
}

interface Step1Props extends StepProps {
  setNextButtonDisabled: React.Dispatch<React.SetStateAction<boolean>>;
}

const Step1: React.FC<Step1Props> = ({
  register,
  control,
  session,
  setNextButtonDisabled,
}) => {
  return (
    <div className="flex flex-col gap-3">
      <Controller
        name={"image"}
        control={control}
        defaultValue={""}
        render={({ field: { onChange, value } }) => (
          <ImageUploader
            image={value}
            setImage={onChange}
            session={session}
            setNextButtonDisabled={setNextButtonDisabled}
          />
        )}
      ></Controller>
      <Input
        label="name"
        type={"text"}
        placeholder="Full Name"
        register={register}
        initialValue={session?.user?.name}
      />
      <Input
        label="portfolio"
        type={"url"}
        placeholder="Portfolio"
        register={register}
      />
    </div>
  );
};

const Step2: React.FC<StepProps> = ({ register, control }) => {
  const { fields, append, remove } = useFieldArray({
    control,
    name: "learningGoals",
  });
  return (
    <div className="flex flex-col gap-2">
      <h3 className="text-p3Med text-myWhite-300">Learning goals</h3>
      <div className="flex flex-col gap-2">
        {fields.map((field, index) => (
          <DeletableListItem
            key={field.id}
            onDelete={() => remove(index)}
            checkable={true}
            register={register}
            fieldArrayName="learningGoals"
            placeholderText="Enter a learning goal"
            control={control}
            index={index}
          />
        ))}
      </div>
      <Button
        image="icons/plus-blue.svg"
        backgroundColor="bg-myBlack-600"
        textColor="text-myWhite-100"
        onClick={() => {
          append({ value: "", completed: false });
        }}
      >
        Add goal checkbox
      </Button>
    </div>
  );
};

const Step3: React.FC<StepProps> = ({ register, control }) => {
  const { fields, append, remove } = useFieldArray({
    control,
    name: "knowledgeLevels",
  });
  return (
    <div className="flex flex-col gap-2">
      <h3 className="text-p3Med text-myWhite-300">Knowledge Level</h3>
      <div className="flex flex-col gap-2">
        {fields.map((field, index) => (
          <DeletableListItem
            key={field.id}
            onDelete={() => remove(index)}
            checkable={false}
            register={register}
            fieldArrayName="knowledgeLevels"
            placeholderText="Enter a knowledge level"
            control={control}
            index={index}
            imageSrc="icons/check-square.svg"
            imageAlt="Checkmark"
          />
        ))}
      </div>
      <Button
        image="icons/plus-blue.svg"
        backgroundColor="bg-myBlack-600"
        textColor="text-myWhite-100"
        onClick={() => {
          append({ value: "" });
        }}
      >
        Add knowledge checkmark
      </Button>
      <div className="flex flex-col gap-2">
        <label
          className="text-p3Med capitalize text-myWhite-300"
          htmlFor={"techStack"}
        >
          {"Tech Stack"}
        </label>
        <TechStackSelector control={control} />
      </div>
    </div>
  );
};

const Step4: React.FC<StepProps> = ({ register, control }) => {
  return (
    <div className="flex flex-col gap-2">
      <div className="flex items-center gap-2">
        <input
          className="rounded-sm bg-myBlack-700 text-myGreen-400 ring-0 focus:ring-0"
          type="checkbox"
          {...register("availability")}
          defaultChecked
        />
        <label className="text-p3Med text-myWhite-300" htmlFor="availability">
          Are you available for a new project?
        </label>
      </div>
      <div className="flex gap-2">
        <MyDatePicker
          label="Start Date & Time "
          fieldName="startDate"
          control={control}
        />
        <MyDatePicker
          label="End Date & Time "
          fieldName="endDate"
          control={control}
        />
      </div>
    </div>
  );
};

export { Step1, Step2, Step3, Step4 };
```

```js
// OnboardingForm.tsx
import { Step1, Step2, Step3, Step4 } from "./Steps";
....
```

dalbanhi commented:
> This feels like it works great- I do have a question about enforcing the square image though. Is there a particular reason you chose to do that?
> 

I chose to do that because I was having trouble getting the image to be properly contained with tailwind -- I think I fixed it however after I had enforced the square image and then never actually tested uploading a square image and didn't realize it broke it to throw an error in the uploadthing callback function. I removed it and did all the other fixes you mentioned.

dalbanhi committed:
warning fix and square image enforcement removed; plus some code cleanup

brandonetter reviewed (approved):
Looks great.

Looks like I'm able to go through with the flow without breaking if I try different aspect ratio of image- and logging back in seems to have me avoid the onboarding like intended :)
